### ISSUE TITLE
Power Kits & Power Hub

### ISSUE THREAD

**Issue by ErikBussink** [tolwi/hassio-ecoflow-cloud#131]
Thanks to this amazing project, I have access to all my data for my Delta Pro & Extra Battery and the PowerStream.  

I also have a Power Kit composed of a PowerHub and 3 LFP Batteries 5kWh. I've run the diagnostics against it and I will attempt to start the support for this Device and Batteries. I'm not a developer, just a tinkerer so bear with me on this.

**Comment by paule96** [comment]
Hi @ErikBussink,

can I somehow help you. I have the same problem / situation.
So if you need help with programming something or some data I can try to provide it.

**Comment by lprhodes** [comment]
I'll be adding this over the next 24h

**Comment by lprhodes** [comment]
The crux of it is working but I still have some work to do tomorrow to ensure all sensors and options are added.

**Comment by lprhodes** [comment]
I find it interesting that with the data we can see exactly how much power is going to the BMS and voltage conversions

**Comment by foxthefox** [comment]
> The crux of it is working but I still have some work to do tomorrow to ensure all sensors and options are added.

can you post the JSON string for the latestQuotas or if not identical to the cyclic updates, them also the JSON of the cyclic response?

**Comment by paule96** [comment]
@lprhodes I started now working too on the support for the power kit. 

I used some of your code for a quick start. My work in progress can be viewed here:

https://github.com/paule96/hassio-ecoflow-cloud/tree/powerkit

**Comment by foxthefox** [comment]
@paule96 can you post a JSON containing latestQuotas and the cyclic telegrams from power kit?

**Comment by paule96** [comment]
@foxthefox there seems to be an update of the API since this weekend. So before this weekend the json was structure like the following:

```json
{
    "code": "0",
    "message": "Success",
    "data": {
        "bbcout": "{some sub json}",
        "lddc": "{some sub json}",
        "bp5000": "{some sub json}",
        "bmsTotal": "{some sub json}",
        "ldac": "{some sub json}",
        "onLineModuleSnList": "{some sub json}",
        "kitscc": "{some sub json}",
        "wireless": "{some sub json}",
        "bbcin": "{some sub json}",
        "ichigh": "{some sub json}",
        "iclow": "{some sub json}"
    },
    "eagleEyeTraceId": "xxxxx",
    "tid": ""
}
```

each subjson was a string usually with an JSON Object that contains the sub devices as Ids. So for example the content of bbcout could look like this:

```json
{
   "SN-OF-SUBDEVICE": {
                "dsgEnergy": 78219,
                "ldOutCurr": 3625,
                "ldOutWatts": 95,
                "workMode": 2,
                "hs1Temp": 39,
                "dcOutSta": 1,
                "l1Curr": 1349,
                "batCurr": 1531,
                "warnCode": 0,
                "batWatts": 77,
                "eventCode": 0,
                "l2Curr": 1430,
                "cfgVolTag": 1,
                "pcbTemp": 0,
                "errCode": 15000,
                "hs2Temp": 0,
                "standbyTime": 0,
                "batVol": 50481,
                "ldOutVol": 26408,
                "dayEnergy": 12343
            }
}
```


Now the response looks like:

```json
{
    "code": "0",
    "message": "Success",
    "data": {
        "bbcout": {
            "snofdevice": {
                "dsgEnergy": 78219,
                "ldOutCurr": 3625,
                "ldOutWatts": 95,
                "workMode": 2,
                "hs1Temp": 39,
                "dcOutSta": 1,
                "l1Curr": 1349,
                "batCurr": 1531,
                "warnCode": 0,
                "batWatts": 77,
                "eventCode": 0,
                "l2Curr": 1430,
                "cfgVolTag": 1,
                "pcbTemp": 0,
                "errCode": 15000,
                "hs2Temp": 0,
                "standbyTime": 0,
                "batVol": 50481,
                "ldOutVol": 26408,
                "dayEnergy": 12343
            }
        },
        "lddc": {
            "snofdevice": {
                "dcTemp1": 27,
                "dcTemp2": 24,
                "dcChRelay": 33,
                "dcChCur": [
                    3676,
                    0,
                    0,
                    0,
                    0,
                    141,
                    0,
                    0,
                    0,
                    0,
                    0,
                    381
                ],
                "dcChSta": 4065,
                "errorCodeAdd": [
                    40000,
                    40040,
                    40080,
                    40120,
                    40160,
                    40200,
                    40240,
                    40280,
                    40320,
                    40360,
                    40400,
                    40440,
                    40800
                ],
                "dcTotalWatts": 110,
                "dcSetChSta": 0,
                "dcInVol": 26664,
                "dcChWatt": [
                    97,
                    0,
                    0,
                    0,
                    0,
                    3,
                    0,
                    0,
                    0,
                    0,
                    0,
                    10
                ]
            }
        },
        "bp5000": {
            "snofdevice": {
                "chgDsgMosState": 3,
                "soc": 5,
                "ptcHeatingEvent": 6,
                "minCellVol": 3178,
                "upsFlag": 0,
                "warnCode": 0,
                "lcdOffConfirmS": 300,
                "ptcChgErrCnt": 0,
                "oilStopUpline": 80,
                "remainCap": 4975,
                "chgState": 1,
                "acDcLsplShutdMin": 0,
                "canId": 1,
                "maxCellTemp": 22,
                "lcdStandbyMin": 300,
                "errCode": 50000,
                "minPtcTemp": 21,
                "totalAmp": 65522,
                "warningEvent": 0,
                "designCap": 100000,
                "kitNum": 1,
                "oilCloseSoc": 80,
                "totalChgDsgState": 1,
                "totalFullCap": 94,
                "amp": -1533,
                "fullCap": 94831,
                "totalOutWatts": -75,
                "ptcRemainTime": 0,
                "totalInWatts": 0,
                "vol": 50921,
                "openBmsIdex": 1,
                "ptcHeatingFlag": 0,
                "remindDsgPtcFlag": 0,
                "minCellTemp": 21,
                "inWatts": 0,
                "proChgDsgMosState": 3,
                "temp": 21,
                "bmsChgUpline": 100,
                "maxMosTemp": 21,
                "doubleOilErrorFlag": 0,
                "remainTime": 181,
                "ptcMosErr": 0,
                "minMosTemp": 21,
                "ptcAllowFlag": 0,
                "maxPtcTemp": 22,
                "eventCode": 50500,
                "maxCellVol": 3191,
                "totalSoc": 5,
                "dsgSetSoc": 0,
                "ptcTouchFlag": 0,
                "balanceFlag": 0,
                "totalRemainTime": 185,
                "bmsFault": 0,
                "oilStartDownline": 20,
                "bmsType": 0,
                "chgSetSoc": 100,
                "oilOpenSoc": 20,
                "outWatts": -78,
                "bmsDsgDownline": 0
            }
        },
        "bmsTotal": {
            "totalChgDsgState": 1,
            "bmsChgUpline": 100,
            "totalFullCap": 94,
            "totalOutWatts": -75,
            "doubleOilErrorFlag": 0,
            "acDcLsplShutdMin": 0,
            "totalInWatts": 0,
            "totalSoc": 5,
            "lcdOffConfirmS": 300,
            "remindDsgPtcFlag": 0,
            "totalAmp": 65522,
            "totalRemainTime": 185,
            "warningEvent": 0,
            "oilStartDownline": 20,
            "oilStopUpline": 80,
            "bmsDsgDownline": 0
        },
        "ldac": {
            "snofdevice": {
                "acTemp1": 31,
                "acChWatt": [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0
                ],
                "errorCodeAdd": [
                    30000,
                    30040,
                    30080,
                    30120,
                    30160,
                    30200,
                    30800
                ],
                "acChCur": [
                    538,
                    0,
                    0,
                    0,
                    0,
                    0
                ],
                "acChSta": 1,
                "acTemp2": 33,
                "acInVol": 233058,
                "acTotalWatts": 6,
                "acSetChSta": 0
            }
        },
        "onLineModuleSnList": [
            {
                "customData": "0",
                "loaderVersion": "1.2.2.2",
                "moduleAddr": "2",
                "moduleDAddr": "1",
                "moduleDetail": "5",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "5.3.5.153"
            },
            {
                "customData": "0",
                "loaderVersion": "1.1.2.2",
                "moduleAddr": "4",
                "moduleDAddr": "1",
                "moduleDetail": "6",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "5.3.4.212"
            },
            {
                "customData": "0",
                "loaderVersion": "4.0.2.2",
                "moduleAddr": "84",
                "moduleDAddr": "1",
                "moduleDetail": "8",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "0.3.0.51"
            },
            {
                "customData": "0",
                "loaderVersion": "2.0.2.1",
                "moduleAddr": "81",
                "moduleDAddr": "1",
                "moduleDetail": "3",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "0.3.1.79"
            },
            {
                "customData": "1",
                "loaderVersion": "1.0.1.15",
                "moduleAddr": "3",
                "moduleDAddr": "1",
                "moduleDetail": "1",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "1.1.2.68"
            },
            {
                "customData": "0",
                "loaderVersion": "3.0.2.2",
                "moduleAddr": "83",
                "moduleDAddr": "1",
                "moduleDetail": "7",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "0.3.0.52"
            },
            {
                "customData": "0",
                "loaderVersion": "2.0.2.1",
                "moduleAddr": "80",
                "moduleDAddr": "1",
                "moduleDetail": "2",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "0.3.0.199"
            },
            {
                "customData": "0",
                "loaderVersion": "1.0.2.2",
                "moduleAddr": "5",
                "moduleDAddr": "1",
                "moduleDetail": "4",
                "moduleSn": "snofdevice",
                "moduleType": "60",
                "moduleVersion": "0.3.0.163"
            }
        ],
        "kitscc": {
            "snofdevice": {
                "pv2ErrCode": 0,
                "pv1InWatts": 0,
                "pv2InWatts": 45,
                "pv1ErrCode": 0,
                "chgEnergy": 0,
                "l1Curr": 0,
                "alt2CableUnit": 0,
                "batCurr": 899,
                "batWatts": 45604,
                "pv1InCurr": 0,
                "pv2InVol": 62076,
                "alt1VoltLmtEn": 1,
                "pv2WorkMode": 1,
                "alt1CableLen": 2000,
                "hs2Temp": 37,
                "mppt2SwSta": 1,
                "alt2CableLen": 600,
                "alt2VoltLmtEn": 1,
                "pv1_hot_out": 1,
                "pv2_hot_out": 0,
                "dayEnergy": 0,
                "mppt1SwSta": 0,
                "dsgEnergy": 0,
                "alt1VoltLmt": 303,
                "alt2VoltLmt": 123,
                "hs1Temp": 35,
                "pv2InputFlag": 1,
                "pv1InputFlag": 0,
                "pv2InCurr": 736,
                "pv1WorkMode": 1,
                "l2Curr": 826,
                "eventCode2": 0,
                "pcbTemp": 0,
                "eventCode1": 0,
                "batVol": 50869,
                "pv1InVol": 0,
                "alt1CableUnit": 0,
                "warnCode2": 0,
                "warnCode1": 0
            }
        },
        "wireless": {
            "snofdevice": {
                "scenes": 4
            }
        },
        "bbcin": {
            "snofdevice": {
                "chgPause": 1,
                "workMode": 1,
                "l1Curr": 28,
                "bpOnlinePos": 1,
                "altCableUnit": 0,
                "batCurr": 78,
                "warnCode": 0,
                "batWatts": 3,
                "altCableLen": 900,
                "altVoltLmt": 123,
                "altVoltLmtEn": 1,
                "hs2Temp": 31,
                "dcInState": 0,
                "workMode2": 0,
                "dayEnergy": 1234,
                "dsgEnergy": 789,
                "inHwTpe": 0,
                "allowDsgOn": 1,
                "hs1Temp": 29,
                "dcInWatts": 0,
                "dcInCurr": 0,
                "chgType": 0,
                "shakeCtrlDisable": 1,
                "eventCode": 0,
                "l2Curr": 26,
                "chargeMode": 0,
                "pcbTemp": 0,
                "errCode": 10000,
                "batVol": 50013,
                "dcInVol": 0,
                "chgMaxCurr": 60000,
                "isCarMoving": 1
            }
        },
        "ichigh": {
            "snofdevice": {
                "outAmp2": 0,
                "invType": 0,
                "inFreq": 0,
                "invSwSta": 1,
                "appCfg": {
                    "powerOn": 1,
                    "wakeup": 0,
                    "standbyTime": 0,
                    "acRlyCtrlDisable": 1,
                    "acFreqSet": 50,
                    "acMaxCurrSer": 2,
                    "acVolSet": 230,
                    "acChgDisable": 1,
                    "passByMaxCurr": 14
                },
                "outVa": 155,
                "inCurr": 0,
                "inVol": 0,
                "outVol": 229982,
                "acTemp": 49,
                "standbyTime": 30,
                "outFreq": 50,
                "cfgOutFreq": 50,
                "outputWhInDay": 0,
                "inputWhInDay": 0,
                "outCurr": 786,
                "ch2Watt": 21,
                "outWatts": 40,
                "inWatts": 0
            }
        },
        "iclow": {
            "snofdevice": {
                "bmsChgCurr": 291,
                "realSoc": 5,
                "warn_code": 0,
                "protectState": 0,
                "chgType": 1,
                "batCurr": 292,
                "dcTemp": 47,
                "chgDsgState": 1,
                "extKitType": 0,
                "busVol": 388442,
                "lsplFlag": 0,
                "maxChgCurr": 0,
                "errCode": 0,
                "event_code": 0,
                "chrgFlag": 1,
                "batVol": 0,
                "fanLevel": 2,
                "chgInType": 1,
                "chgBatVol": 50910
            }
        }
    },
    "eagleEyeTraceId": "xxxx",
    "tid": ""
}
```

> Note: Everywhere where `snofdevice` is written in the json, the serial number of the subdevice is in the original JSON there. (for example from the battery)

**Comment by foxthefox** [comment]
thank you very much!

again a different approach in data structure, but luckily still JSON

Maybe you the answers to some questions:
- is there a latestQuotas cmd? Most likely the one causing the complete reply, but at other devices the JSON starts for latestQuotas like:
```
{
	"id": 141391078,
	"moduleType": 0,
	"operateType": "latestQuotas",
	"version": "1.1",
	"data": {
		"sn": "*********",
		"productType": 80,
		"model": 1,
		"online": 1,
		"quotaMap": {
                   ...
		}
	}
}
```
- the telegrams are received on the serial number of the kit and the subdevices do not need a subscription?

**Comment by tolwi** [comment]
@paule96 take a look at [this](https://github.com/tolwi/hassio-ecoflow-cloud/compare/140-new-device-smart-home-panel) branch - it is abandoned, but you can catch the way on how to [handle](https://github.com/tolwi/hassio-ecoflow-cloud/blob/140-new-device-smart-home-panel/custom_components/ecoflow_cloud/entities/__init__.py#L50) nested json

**Comment by foxthefox** [comment]
@paule96 looking to the complete JSON, the bmsTotal has no "snofdevice", thats not a problem, but it is different to the others. Looking to your powerkit.py I would assume that the incomming JSON has already be somehow tweaked to get rid of the intermediate "snofdevice", otherwise the values most likely are not catched with the expresssion.
Also the commands seem to be unchanged from a powertstation ('pd' is not part of the powerkit), the commands somehow not so clear for me.

**Comment by paule96** [comment]
@tolwi and @foxthefox thanks for the input. How can I produce more than one device per integration entry?
![image](https://github.com/tolwi/hassio-ecoflow-cloud/assets/11291885/ba54d5ef-8280-435e-824c-5fca3a1725b7)

Because in my understanding, the powerhub of the powerkit is more like a gateway that manages other devices, like the screen, like the smartdistribution panel, like the batteries, or even another hub (if you use it in combination with the smart panel)

**Comment by paule96** [comment]
@foxthefox the json was unmodified, just the SNs where replaced. And yes it is weird that bmstotal has no sn.
But also like understandable. Because I guess it's like a sum entity over all connected batteries. So it's directly connected to the powerhub

**Comment by foxthefox** [comment]
> @tolwi and @foxthefox thanks for the input. How can I produce more than one device per integration entry? 
> 
> Because in my understanding, the powerhub of the powerkit is more like a gateway that manages other devices, like the screen, like the smartdistribution panel, like the batteries, or even another hub (if you use it in combination with the smart panel)

For the powerkit it seems that there are different "subdevices" which might need to be treated as another devices. From my perspective there is still the same ID of the main device (may be the "gateway") so the "subdevices2 can be treated in the same way as the different portions of a powerstation (bms,pd,ems,mppt,...). At powerstations they do not appear as "subdevices with own ID/moduleSn" but actually they are used in the same way. In any case additional batteries are special. At powerstations they appear as bms..1 or bms..2 and at powerkit it is a next serial# within bp2000/bp5000. So again it is compareable.

The question would be more how the different "channels" could be used to structure the data within the integration. Otherwise all numbers (indpendent from the channel/subdevice) appear in one set. This is more serious when all available data is read into the integration.
As stated already somewhere else, I have still problems to understand the Homeassistant approach of managment of objects and the way how to setup a development environment, so I am not sure if intermediate structures like channel/subdevice is supported or not.

I am working on an ioBroker adapter for EF which in the meantime is also a HA gateway. It uses the MQTT discovery function and there is a hierarchical approach possible.
See some screenshots:

entry with several devices:

![Bildschirmfoto 2024-06-09 um 11 16 53](https://github.com/tolwi/hassio-ecoflow-cloud/assets/16841643/cf61515b-a764-409e-90b5-331166179043)

one device with channels/subdevices:

![Bildschirmfoto 2024-06-09 um 11 18 22](https://github.com/tolwi/hassio-ecoflow-cloud/assets/16841643/1116f570-5f07-4c07-8c93-7647162a7302)

the channel/subdevice:

![Bildschirmfoto 2024-06-09 um 11 19 44](https://github.com/tolwi/hassio-ecoflow-cloud/assets/16841643/0d9f9ba0-3d9b-41af-a1c0-c945de89b4aa)


maybe it gives an idea for the integration. But I am not sure if it is supported at all, when using the native ways (instead of MQTT).

**Comment by paule96** [comment]
hm okay. if we are not sure if it's supported maybe we should then just provide each devices as it's own root device. From the ecoflow mqtt documentation for the powerkit, it is possible to query each device by itself. For example, for the bbcin it is:

```json
{ "sn": "M106ZAB4Z000001F", "params": { "quotas": [ "bbcin" ] } }
```

Then the question is, how todo it in the registration / config step. 
Currently my understanding of this integration is that in the following lines of code, you get the devices implmentation and register it as an homeassistant device:

```python
from .devices.registry import devices
            device = devices[user_input[CONF_DEVICE_TYPE]]

            options = {OPTS_POWER_STEP: device.charging_power_step(), OPTS_REFRESH_PERIOD_SEC: DEFAULT_REFRESH_PERIOD_SEC}

            return self.async_create_entry(title=user_input[const.CONF_NAME], data=user_input, options=options)

```

But it would be nice if `devices would be just a list`. But then the problem is, how  to add the SN of the subdevices and the category (like 'bbcin').

> P.S.: @foxthefox  I'm still struggling with how this integration works at all. If you have time today, it would be nice to have a hands-on call later. (Right now I'm in German train, but in around 2h to 3h I should be available, for example a teams call)

**Comment by foxthefox** [comment]
Ich weiß leider nicht wie man einen privaten Chat über git machen kann, aber wenn es keine Umstände macht, dann können wir im https://forum.iobroker.net/ Infos austauschen, dort auch als foxthefox.

**Comment by paule96** [comment]
to just give a little status update to all:
I'm now able to ready all the data from a power kit with a single battery and a distribution panel connected to it.

The next steps are to implement multi battery support and most of the switches. Then I will be confident to create a pull request.

But right now you here you can see some of the data it is able to produce:
![image](https://github.com/user-attachments/assets/1aeea69c-b2c5-4530-8ef0-58ba8ea87317)

**Comment by diablo581** [comment]
Has there been any update on this? I would also like support for this if possible.

**Comment by foxthefox** [comment]
When working on the iobroker adapter which uses the unofficial API I had discovered that even there is an incomming JSON object the specific Objects like bbcout/bbcin... are wrapped additionally into a string. So you need to treat the message and it parts in a special way. I have solved that with version 1.04 of my adapter.
Beside of that discovery there are also update telegrams which contain hourly energy values, which probably are not part of the official API update.

**Comment by paule96** [comment]
@diablo581 I'm close to make a pull request to this repository. I fight a lot with the python stuff lately and most of the stuff works in theory :) 

Need to cleanup some code.

If you want to see the current state of it you can take a look [here](https://github.com/paule96/hassio-ecoflow-cloud/tree/powerkit).

**Comment by diablo581** [comment]
@paule96 attempting to use your branch, but I don't see powerkit as a selectable model when I go to set it up.  What do you select to bring it in?

**Comment by paule96** [comment]
@diablo581 sorry for the late response upfront. I was sick the last days. 

But to help you it works as following:

1. clone the repository and switch to branch `powerkit`
  - I recommend cloning into dev container in VsCode
  - alternative after cloning start the dev container
1. Starting the devcontainer will take a lot of time. (and a lot of data will be transfer over the internet)
1. If everything ready you can lunch the debug config `HomeAssistant` (for the first start)
1. then after a very long first start you can navigate on your local PC to `http://127.0.0.1:8123/` an complete the HA setup
1. now you can configure the ecoflow integration (it's the old configuration setup not the new one)
1. Then you should be able to use your powerkit in this dev instance

Problems with the current way of development:

- breakpoints are not working, because the source will be copied to homeassistant
  - to debug use the files in `core/config/custom_components/ecoflow_cloud`
- after the first start of the dev instance use the task `Home Assistant (skip pip)` to be faster


I know that this is not perfect yet, so if anybody as ideas to tell the python debugger that the src files are somewhere else that would be really helpful for this #248 


I'm soon in my winter break and will have hopefully time to update the branch again and also apply the new and way better configuration experience.

**Comment by paule96** [comment]
Again a small update from my side.
I was able to merge the main branch into my changes and can now manage all the sub devices of a powerkit separate by each other:
<img width="1671" alt="image" src="https://github.com/user-attachments/assets/42f64343-27c0-4b3c-a7da-d2ba8639386e" />


The red lines are the serial numbers in the format of:

`hub.subdevice`

So it looks like:

`M109xxxxxxxxx.M101xxxxxxx` 

For a battery for example. (I didn't sound the `x`s so maybe more or less characters are needed)

**Comment by paule96** [comment]
@diablo581 I change the main branch in my repo to the powerkit branch. With this change you should be able to try it out with a normal home assistant installation and HACS. 
Just add my repo and update the ecoflow cloud integration. (https://github.com/paule96/hassio-ecoflow-cloud.git)

But I highly recommend a backup before. 
(I tried the migration of the old config items with my ecoflow devices. They worked so far...)

**Comment by diablo581** [comment]
Oh excellent! I'm going to mess with it this morning.

I don't have any other ecoflow stuff, so if I can get this to work then I will be golden.

I appreciate you working on this!

**Comment by diablo581** [comment]
So I still don't see powerkit listed as an option.  I removed all EcoFlow repositories, and readded yours as a custom repository.  Added the integration via HACS and restarted HA.  Then I went in, told it I wanted to add something via the EcoFlow-Cloud integration, and used my account to login.  Then when I get the drop down for the device I don't see powerkit listed anywhere.

**Comment by diablo581** [comment]
Wait! Looking at the code it looks like it is only a selectable option when you use the public API.  That explains why I can't see it when I use my account.  I'm not sure exactly how you use the public API though

**Comment by diablo581** [comment]
Sorry to spam. I looked up the instructions for getting the API and Secret keys for the public API. I have submitted my account for developer access to get it it.

**Comment by paule96** [comment]
@diablo581 yeah you need to use the public api. I can try to add it to the other on too. But be aware that the other way is not support by ecoflow and can result in an account block. (not yet happend to anyone as dar as I know )

**Comment by paule96** [comment]
@diablo581 I have bad news here. It is not possible to add the functionality to the private api because the private api has no `quoata/all` endpoint. Without this I don't see an easy way to query all the subdevices of the powerkit. 

On the other side I did write ecoflow an email, if their is a way for integrations like this one to integrate better into the normal login experiance. (Like we know from many other services, where you can login via google, facebook, microsoft, and still use this 3rd party service)
Would be nice to have that, instead of the current id/secret stuff an enduser must do right now.

**Comment by diablo581** [comment]
OK got my API and Secret and couldn't get it to work until I read a different thread where they talked about having to change the base URI to the North American endpoint.  Once I did that I was able to get it connected and pulled in.  Looking through all the stuff now, but so far it has worked! Thank you!

**Comment by paule96** [comment]
@diablo581 by any chance can you write about your hardware configuration? I'm in special interested if you have:

- 2kwh batteries
- multiple batteries (any configuration)
- an alternator connector

Maybe you can share some Screenshots of it how it is shown with my code. Because these stuff I can't test with my setup.

But nice to hear that it works for you now too.

**Comment by diablo581** [comment]
![image](https://github.com/user-attachments/assets/8531fc7d-4866-4246-b6a5-bdc56067e297)

So I have two of the bigger batteries. They show up like this.  I don't have an alternator connected, but I do have solar and AC connected to the system.  What else would you like to see?

**Comment by diablo581** [comment]
Yep. They report different temps, voltages, etc.  I am going to build an automation that turns the heat off out in this location as long as the batteries are above 40 degrees.  If they dip below then it will turn it back on.

**Comment by JustEnf** [comment]
I have configured my Power Kit (Gen 1) in v1.3.0 integration except that the PK does not show up as a Device/Entities in HA

**Comment by paule96** [comment]
@JustEnf the current implementation is broken. (after some commits in the beta)
Theirs is a draft pull request open that fixes that but it has quirks that's why it is a draft.
@tolwi should left a comment on the draft PR or approve the bigger changes. Until now the powerkit only works with an older beta or with the draft PR. (or with my repository that has a default branch what mateches the old beta)

**Comment by diablo581** [comment]
Yep I updated mine from your branch hoping that all the fixes where in and mine is broken again. I'm going to have to revert back.

**Comment by diablo581** [comment]
Has anyone tested the 1.30 release to see if it works with t his?  I've stayed with paule96's verson so far.

**Comment by micradot** [comment]
Same issue with my PowerKit gen2 50A

Device: PowerKit
API: Public API (api-a.ecoflow.com)
HA: 2025.11.3 (Home Assistant OS)
I tried the following integrations:
tolwi version: 1.4.0-beta3
tolwi version: 1.3

Error from logs:

> custom_components.ecoflow_cloud.api.EcoflowException: current device is not allowed to get device info

The API successfully authenticates and can see the device name and serial number during setup, but fails when trying to retrieve device data. I've also submitted a support ticket to EcoFlow directly.

@paule96  - I saw your fork is recommended for PowerKit. I'm trying to install it via HACS but having trouble finding it. Is your repo still the best option for PowerKit support?

**Comment by paule96** [comment]
@micradot yeah it should just work. And you can not find it. You need to add the repo manually to hacs.

**Comment by micradot** [comment]
@paule96 

Update from EcoFlow support:
“Currently, the Power Hub 5KVA does not support API access. Only the first-generation Powerkits support this feature.”

So Gen2 PowerKit is explicitly unsupported on their end - not an integration bug. Asked them about timeline and local alternatives, will update if they respond.

**Comment by JustEnf** [comment]
I'm back to try to get the latest v1.4.1 to work with my Gen1 Power Kit. Currently I only see one 'Control' named "Reconnect' showing on the device's details page. I don't see any other Entities for the device.

I'm not sure where to go from here to see the Entities. I'm using the US API.

**Comment by diablo581** [comment]
This is all I saw as well. I rolled back to @paule96 tree.

**Comment by JustEnf** [comment]
Diablo581 How does one do that?

**Comment by diablo581** [comment]
Click his name up there and it should take you to his repo.

Then you can add it to HACS and remove the one from this repo and then download his.

**Comment by JustEnf** [comment]
diiablo581: Can I contact you off this thread? I did the steps and they went fine. However I get a Migration Error in the Integration.. tried to re-add the Power Hub and errors... Someday.....

**Comment by diablo581** [comment]
I think you may still be running the other one. I've only gotten migration errors with the original one.

I'd delete everything, uninstall everything, remove this repo, then restart HA and start "fresh".

If that doesn't fix it I'm at a loss, but these purge five gentlemen might be if help if you post a log.

**Comment by JustEnf** [comment]
<img width="570" height="377" alt="Image" src="https://github.com/user-attachments/assets/490e48da-d80a-45db-a6ad-119d5659ac8c" />

<img width="688" height="643" alt="Image" src="https://github.com/user-attachments/assets/5c065e69-6578-4893-8ad2-37b0525ce2fb" />

I have now deleted everything and started over. so no Migration error but still the same error as before.  'Power Hub' is in the list of my 4 devices but I get the following error because 'Power Kits' is not in the of 'Device Type' list, blocking me from adding it. I can add the other devices without error, just not the Power Hub. 

'value must be one of ['DELTA 2', 'DELTA 2 MAX', ........ etc'   If the Power Hub is 'allowed' to be in my list of returned devices then why is 'Power Kits' not in the Device Types list?

Any help from anyone would be appreciated. I have spent hours on educating myself and trying to get this work for my Power Hub (Gen 1).  It is the whole reason I started this journey. The unit is in my campervan so it is my whole reason form doing this. I am at my mental ends but I really need this.

I'm using v1.1.7 integration

**Comment by diablo581** [comment]
Make sure you are using this one, because I don't think that version number is right

https://github.com/paule96/hassio-ecoflow-cloud 

I have Version 1.3.0-beta4 from @paule96

**Comment by JustEnf** [comment]
@diablo581 Thank you! It's working. After banging my head forever... One caveat, this Integration doesn't support the Wave 2 or Glacier and I can't use the one that does, side by side, since they share the same HA folder. At least I have my Power Hub and Delta 2 available.
